File: p021/p021.py
import math
def sumOfDivisors(n):
    """
    A function that returns the sum of proper divisors of a given number n.
    """
    if n == 1:
        return 0
    dSum = 1
    sqrt = math.sqrt(n)
    rsqrt = int(sqrt)
    for i in range(2, rsqrt):
        if n % i == 0:
            dSum += i + n / i
    if sqrt == rsqrt:
        dSum += sqrt
    return int(dSum)

def sumAmicableNum(n):
    d = []
    amiSum = 0
    for i in range(n):
        d.append(sumOfDivisors(i))
    for i in range(2, n):
        dSum = d[i]
        if dSum < n:
            if d[dSum] == i and i != dSum:
                amiSum += dSum
    return amiSum
        
if __name__ == "__main__":
    print(sumAmicableNum(10000))

File: p021/p021.js
function sumOfDivisors(n) {
  if (n == 1) return 0;
  let sum = 1;
  let sqrt = Math.sqrt(n);
  for (let i = 2; i < sqrt; i++) {
    if (n % i == 0) {
      sum += i + n / i;
    }
  }
  if (Math.floor(sqrt) === sqrt) {
    sum += sqrt;
  }
  return sum;
}

function sumAmicableNum(n) {
  // Good luck!
  const d = Array(n);
  let sum = 0;
  for (let i = 1; i < n; i++) {
    d[i] = sumOfDivisors(i);
  }
  for (let i = 2; i < n; i++) {
    let dSum = d[i];
    if (d[dSum] === i && i !== dSum) sum += dSum;
  }
  return sum;
}
console.log(sumAmicableNum(5000));
